### DISCUSSION TITLE
GPS vehicle track logging with ESP8266

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#17955] [Show and tell]
Initial try of a vehicle-based GPS logger using Tasmota. Using a cheap [GPS module](https://www.aliexpress.us/item/3256802336701002.html) from Aliexpress hooked up to ESP8266 compiled with 
```
#undef USE_RULES

 #define USE_SCRIPT 
  #define SUPPORT_MQTT_EVENT                
 #define USE_SCRIPT_WEB_DISPLAY
 #define USE_SCRIPT_JSON_EXPORT
 #define USE_SCRIPT_TIMER
 #define USE_SCRIPT_SPI
#define TS_FLOAT double //note that this is a fork from @gemu2015 that hasn't been merged yet
#define USE_UFILESYS
#define USE_GPS
```
Also `#undef` as much other stuff as possible because RAM is close to the edge with current defines. System is running at about 14k heap. Tried with `#define USE_SDCARD` but kept getting exceptions.

platform_override.ini is adjusted to:
```
board                   = esp8266_4M3M
board_build.f_cpu       = 160000000L
```
to take advantage of the 3MB flash space not used for firmware.

Script is as follows (using gemu's scripteditor to clear out comments and load directly):
```
>D 48
SB=4096
IP=192.168.4.1
lat=0
lon=0

;     multiply lat/lon by 1million to make math easier, maybe not necessary if using TS_FLOAT double
latnum=0
lonnum=0

;     Horizontal accuracy, don't care about vertical accuracy
hacc=0

;     For speed calculations, u-blox messages already contain it, but not implemented in driver
oldlat=0
oldlon=0

;     oldlat multiplied by 1million
boldlat=0
boldlon=0

;     not sure about nested math operations, so declaring extra variables to separate them
latspd=0
lonspd=0
oldspd=0
spd=0

cnt=0
;     altitude, not really necessary
;alt=0

res=0
fr=0

;     buffer counter that fills when movement is happening
ehold=0
newpos=0
cntr=0
dcml=0
tmp=0

str=""
string="\n"

>B
; set sensor file download link
;fl1("truck.txt")
; delete file in case we want to start fresh
;fd("truck.txt")
res=is(170 "")

; list all files in root directory
fr=fo("/" 0)
for cnt 1 20 1
res=fr(str fr)
if res>0
then
print %cnt% : %str%
else
break
endif
next
fc(fr)

;    Define D4/GPIO2 (D1 Mini LED) as output
spinm(2 1)

>T
;     get sensor values, assigning oldlat value doesn't work here
;oldlat=lat
;oldlon=lon
lat=GPS#lat
lon=GPS#lon
alt=GPS#alt
hacc=GPS#hAcc
newpos=GPS#newpos

>S
;     implement drift filter outside of FLOG functionality. Easier to just implement in script code
;print %timer1%
;if timer1==0 {
;timer1=-1
;=>sensor60 2
;}

;     Reset string to prep for each write
str=""

;     Calculate speed based on lat/lon movement pythagorean root(latdiff^2+londiff^2) and change meters/sec to km/h. Using 6 digits of precision to allow reaction to mimimum GPS accuracy of 1.11m (5-digits)
latnum=int(lat*1000000)
lonnum=abs(int(lon*1000000))
boldlat=int(oldlat*1000000)
boldlon=abs(int(oldlon*1000000))
latspd=(boldlat-latnum)*.4329
latspd*=latspd
lonspd=(boldlon-lonnum)*.4329
lonspd*=lonspd
spd=lonspd+latspd
spd=pow(spd 0.5)
;     Debug print
print spd=%2spd% oldspd=%2oldspd% 
;lat=%6lat% lon= %6lon% ehold=%0ehold% heap=%0heap% hacc=%2hacc% tmp=%0tmp%

;     If fix exists (lat>0) and speed greater than 1km/h (walking speed is ~6km/h and drift is usually less than 1, then write new position to the buffer array)
if (spd>1 and lat>0) {
str+=hx(epoch)
str+=hx(latnum)
str+=hx(lonnum)
str+="\n"
ehold+=1
is[ehold]=str
;     Debug print
;print %str%,%0newpos%,%0ehold%,%0latnum%,%3hacc%,%0lonnum%,%0epoch%,%0heap%
}

;     If buffer is full or truck has just stopped, (i.e. speed is dropped below 1 and previous speed was greater than 1) then write the buffer to flash. Usually takes a few seconds to settle, should be enough time to do a write before power is cut off. Also filter weird high values.
if ((ehold==170) or (spd<1 and oldspd>1 and oldspd<150)) {
fr=fo("truck.txt" 2)
for cntr 1 ehold 1
res=fw(is[cntr] fr)
next
fc(fr)
ehold=0
print write
}

;    Blink LED when flash space is getting low
fr=fo("truck.txt" 2)
tmp=fz(fr)
fc(fr)
if tmp>2800000 {
spin(2 upsecs)
} else {
spin(2 1)
}

oldspd=spd
oldlat=lat
oldlon=lon

>W
<center>Speed = %spd%<br>

>J
,"Logger":{
"Speed":%spd%
}
>R
```

Notes:  
- Cold start of GPS module takes between 20 seconds and 2 minutes, so future work will include a lithium cell to keep the module running while vehicle is off so location can start logging immediately. Should consume about 30mA, and can be powered directly from a 18650 (GPS module runs off 3.3-5.0v, so will last minimum 60 hours. ESP8266 is unpowered until vehicle is started.

- Writing 24 bytes for each position, which could be cut in half by writing individual bits with `fwb(byte fr)` function, but that got complicated with file formatting and the buffer. Still get 34 hours of driving logs, which for me is at least a month of driving.

To do:
- Clean up user_config_override and attach here
- Submit PR for adjusted GPS driver code
- Package everything nicely
- Try writing bytes directly to file
- Automate file download when in range of home wifi

Comments and suggestions are welcome

**Comment by daexpert** [comment]
Great work, This is exactly what I need. Can you compile this into a bin file so it can be flashed to esp with tasmota? Thank you.

**Reply by pkkrusty** [reply]
I have my phone's hotspot hard coded in, as well as a few other things. I'll see about removing all that stuff and having the initial wifi config page being broadcast. Better give me a day or two. I'll also attach the excel file that decodes the hex data, which can then be pasted into https://www.gpsvisualizer.com to get a nice .kml or .gpx file.

**Reply by pkkrusty** [reply]
[tasmota.bin.gz](https://github.com/arendst/Tasmota/files/10812833/tasmota.bin.gz)
[user_config_override.txt](https://github.com/arendst/Tasmota/files/10812839/user_config_override.txt)

Probably best to adjust the GPS driver code to receive already calculated speed data from the module itself, rather than calculating it, but I haven't dived into that yet. For now, I've added a JSON publish on teleperiod to the script

**Comment by daexpert** [comment]
To be more specific, I'm intrested to measure the car speed in realtime, the esp flashed with tasmota to display the speed and send the value using mqtt protocol to an mqtt srever.

**Reply by barbudor** [reply]
What @pkkrusty designed is a logger, not a tracker. It records the position to the SD card.
You would need some GSM/4G/LTE kind of connectivity to transmit in real-time

**Reply by daexpert** [reply]
That's why tasmota is ideal for my situation, I can make a host spot from my phone and the esp flashed with tasmota connects to the phone and then publishes the data using mqtt to an mqtt server.

**Reply by pkkrusty** [reply]
That's even easier if you don't care about battery life of your phone. Probably would need some back-end work to parse MQTT status messages (which you could send every 10 seconds to a public server that you run, or sign up for one of those that barbudor mentioned. 

But also quite a different project than my logger (not tracker). 

Flespi's free tier would probably work for what you're wanting, but I've never taken the time to work with an external MQTT server, so I' not the person to guide you through that.

**Reply by daexpert** [reply]
I don't have programming skils, only networking skils, so for me it's hard, I love tasmota it's incredibly simple(the web interface) but didn't found any project that uses tasmota to display the speed. Your project was the closest to what I've needed, but now I'm very disappointed that you can't help me.

**Reply by pkkrusty** [reply]
A few tweaks to the script can display speed on the front page of the GUI. I can add that on Wednesday. But the MQTT portion I can't help you with.

**Comment by daexpert** [comment]
That would be great too if you can display the speed, maybe  I'll figure out how to send mqtt after that. Thank you very much.

**Reply by pkkrusty** [reply]
Script is adjusted to add `>W
    <center>Speed = %spd%<br>` near the bottom.

**Reply by daexpert** [reply]
Thank you so much, you're the man, I will test it.

**Comment by daexpert** [comment]
Yes, please send it to me if you have it, as I mentioned I don't have
programming skiills, I didn't manage to compile a bin file using the speed
parameter indicated by you in previous thread.

**Reply by pkkrusty** [reply]
Linked up above. I just pushed a PR to add speed and heading to the driver, so wouldn't need scripting to access this info. Easier to integrate with other services. We'll see if it gets merged...

**Comment by daexpert** [comment]
Tell me if I did something wrong:
1. I've copied your xsns_60_gps.ino file
2. accessed gitpod.io
3. replaced the existing file in the following folder
/workspace/Tasmota/tasmota/tasmota_xsns_sensor with yours
4 compiled it in terminal using command: platformio run -e tasmota
5. downloaded the new firmware
6. the problem is that the new tasmota.bin file has exactly the same size
as the original, I suppose nothing has changed, so I can't find gps_rx and
gps_tx in tasmota configuration module webpage after selecting generic(18)
module type
I think it would be easier if you send me the bin file because as i've said
i have no programming skils

**Reply by pkkrusty** [reply]
Yeah, you probably have to adjust the #defines also.  Here's the latest binary I flashed on my device.
[tasmota.bin.gz](https://github.com/arendst/Tasmota/files/10818559/tasmota.bin.gz)

**Comment by daexpert** [comment]
Thank you so much. It works. It even sends the speed to the mqtt server. The only thing that I have to do now is to find a solution to read the speed in real time, so far I have found that the minimum TelePeriod is 10 and that's not acceptable.

![Screenshot 2023-03-06 213321](https://user-images.githubusercontent.com/78557151/223213308-891cd987-ec71-49d1-b463-36b6d42ff97d.png)

**Reply by nikito7** [reply]
> Thank you so much. It works. It even sends the speed to the mqtt server. The only thing that I have to do now is to find a solution to read the speed in real time, so far I have found that the minimum TelePeriod is 10 and that's not acceptable.
> 
> ![Screenshot 2023-03-06 213321](https://user-images.githubusercontent.com/78557151/223213308-891cd987-ec71-49d1-b463-36b6d42ff97d.png)

With esp8266 less than 10s it will crash.

Can't keep enough free heap

**Reply by sfromis** [reply]
If you can't make sure to have a reasonable amount of free heap, you need to change your build or usage to use less. Alternatively, you'd get more available if using ESP32.

**Comment by Sum-Guy** [comment]
Just wondering if this project is still active.  I'd like to use an ESP module (either 8266 or 32) as a GPS logger.   Something to log GPS coordinates and make minimal or no use of wifi transmissions while logging.  Then I'd connect to a USB port to download the logged data.  Is this possible?

**Reply by pkkrusty** [reply]
It's active inasmuch as I am still using my logger. ESP8266 is purpose-built for wifi, and it sounds like you're looking for a non-wifi solution. I'm sure there are many Arduino-based logger projects out there with comprehensive how-to guides. 

And depending how much your time is worth, it may be worth it to just buy off-the-shelf. AliExpress has lots of logger options for $20-$30.

I built a ESP32 version of this project with a 7-segment display and a push-button so the user could cycle through categories and the logging would track that too (for mileage reimbursement).

**Comment by Sum-Guy** [comment]
Yes, right after I posted that, I thought about an arduino solution instead.  Regarding the loggers for sale (Aliexpress or ebay) - they ALL seem to be cellular service (sim card) based.  And none of their descriptions indicate they will work as a logger with local storage and without a sim card installed.  They don't even seem to have a usb port - to charge or to pull off any log files.  So can you tell me if they do indeed work as just a logger (ie not just a tracker) ?
